### PR TITLE
Add flow affinity groups

### PR DESCRIPTION
**Pull request by dpoltavets**

### CODE REVIEW

File: src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowFrame.java [hunk 1]
```diff
@@ -197,12 +198,20 @@
     public abstract void setProtectedReversePathId(PathId protectedReversePathId);
 
     @Override
-    @Property(GROUP_ID_PROPERTY)
-    public abstract String getGroupId();
+    @Property(DIVERSE_GROUP_ID_PROPERTY)
+    public abstract String getDiverseGroupId();
```
Comment (sergii-iakovenko) [inline src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowFrame.java:202]: Will this work with the blue-green deployment approach? DB migration doesn't guarantee that  when it's done the previous version of the system wont update old fields in DB.

**Review by sergii-iakovenko — CHANGES_REQUESTED** [review]

File: src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowFrame.java (hunk 1 above)
Comment (dpoltavets) [reply src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowFrame.java:203]: This approach will not work with the blue-green deployment. The approach has been reworked. Now GROUP_ID_PROPERTY is used, and we write diverse group id value to GROUP_ID_PROPERTY and DIVERSE_GROUP_ID_PROPERTY at the same time. Second step is starting use DIVERSE_GROUP_ID_PROPERTY (#4404). Third step is removing group id field (#4405).

File: src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowEventDumpFrame.java [hunk 2]
```diff
@@ -186,12 +186,20 @@
     public abstract void setReverseMeterId(MeterId reverseMeterId);
 
     @Override
-    @Property("group_id")
-    public abstract String getGroupId();
+    @Property("diverse_group_id")
```
Comment (sergii-iakovenko) [inline src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowEventDumpFrame.java:189]: Does it mean that you wont be able to read stored group_id field in this release until the DB migration from the next release is done?

Comment (dpoltavets) [reply src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/FlowEventDumpFrame.java]: Fixed.

File: src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/AvailableNetwork.java [hunk 3]
```diff
@@ -165,7 +169,58 @@ public void processDiversitySegmentsWithPop(List<PathSegment> segments) {
...
+        for (Node node : pathNodes) {
+            Collection<Edge> pathNodeEdges =
+                    Stream.concat(node.getOutgoingLinks().stream(), node.getIncomingLinks().stream())
+                            .filter(edge -> !pathEdges.contains(edge))
+                            .collect(Collectors.toList());
+            for (Edge edge : pathNodeEdges) {
+                edge.increaseAffinityGroupUseCounter(affinityGroupUseCounter);
+            }
 
+            affinityGroupUseCounter++;
```
Comment (rozdy) [inline src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/AvailableNetwork.java:216]: Didn't get it. Why do you need to increase counter?

**Review by sergii-iakovenko — APPROVED** [review]

**Review by rozdy — CHANGES_REQUESTED** [review]

File: docker/db-migration/migrations/011-add-diverse-and-affinity-group-id-to-flow-class.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  - changeSet:
+      id: add_diverse_and_affinity_group_id_to_flow_class
+      author: dpoltavets
+      changes:
+        - sql: "CREATE PROPERTY flow.diverse_group_id IF NOT EXISTS STRING"
+        - sql: "CREATE INDEX flow.diverse_group_id NOTUNIQUE_HASH_INDEX"
+        - sql: "CREATE PROPERTY flow.affinity_group_id IF NOT EXISTS STRING"
+        - sql: "CREATE INDEX flow.affinity_group_id NOTUNIQUE_HASH_INDEX"
+        - sql: "CREATE PROPERTY flow_dump.diverse_group_id IF NOT EXISTS STRING"
+        - sql: "CREATE PROPERTY flow_dump.affinity_group_id IF NOT EXISTS STRING"
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docker/db-migration/migrations/011-add-diverse-and-affinity-group-id-to-flow-class.yaml:18]: This query is performed in #4404

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+FlowHSTopology -> OrientDB : get affinity flow by id
+OrientDB --> FlowHSTopology : Flow instance
+alt flow affinity group property is not defined
+   FlowHSTopology -> OrientDB : save affinity flow id as affinity group id
+end
+
+FlowHSTopology -> PathComputerFactory : getPathComputer
+create PathComputer
+PathComputerFactory -> PathComputer : create
+PathComputer --> FlowHSTopology : PathComputer
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml]: Fixed.

File: docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md [hunk 6]
```diff
@@ -34,6 +34,8 @@ Flow groups is an implementation detail and there is no API to access it directl
 
 System knows nothing about physical network topology, so computed paths not truly diverse in hardware meaning.
 
+In the current implementation, affinity and diverse groups can not be specified at the same time for one flow.
+
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-diverse-flows/pce-diverse-flows.md:39]: Crosslinks have been added

File: src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/model/FlowDto.java [hunk 7]
```diff
@@ -234,6 +237,7 @@ public FlowDto() {
      * @param pathComputationStrategy   path computation strategy
      * @param targetPathComputationStrategy   target path computation strategy
      * @param diverseWith               flow ids diverse with
+     * @param affinityWith               flow ids diverse with
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/base-topology/base-messaging/src/main/java/org/openkilda/messaging/model/FlowDto.java]: Fixed.

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+FlowHSTopology -> PathComputer : get path with flow group id
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml]: Fixed.

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+group Build AvailableNetwork
+    AvailableNetworkFactory -> OrientDB: get active links
+    OrientDB--> AvailableNetworkFactory
+    create AvailableNetwork
+    AvailableNetworkFactory -> AvailableNetwork : create from retrieved links
+    AvailableNetwork --> AvailableNetworkFactory
+
+    group Fill network affinity weights
+        AvailableNetworkFactory -> OrientDB : get flow segments for flows with the affinity group id
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml]: Fixed.

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        OrientDB--> AvailableNetworkFactory : flow segments
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml]: Fixed.

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+        AvailableNetworkFactory -> AvailableNetwork: fill affinity weighs using\ndiverse flow segments
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml]: Fixed.

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    PathComputer --> FlowHSTopology : computed path
+    deactivate PathComputer
+end
+
+alt path found
+    FlowHSTopology -> OrientDB : create new flow with affinity group id
+
+    FlowHSTopology --> NB : flow create operation response
+
+else path don't found
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows-create.puml]: Fixed.

File: docs/design/solutions/pce-affinity-flows/pce-affinity-flows.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
+# Affinity flows
+
+## Goals
+The main goal is to be able to search for paths that maximally repeat the path of a certain main flow.
+
+## NB contract changes
+Create\Update Flow Request: add optional flow id, to make flow affinity with.
+
+Flow Response: If flow is in an affinity group, flow response will contain an `affinity_with` field 
+with a list of the flows in that group.
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/pce-affinity-flows/pce-affinity-flows.md]: Now only the id of the main flow is returned

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java [hunk 14]
```diff
@@ -355,6 +361,34 @@ void checkDiverseFlow(RequestedFlow targetFlow) throws InvalidFlowException {
...
+    void checkAffinityFlow(RequestedFlow targetFlow) throws InvalidFlowException {
+        if (targetFlow.getSrcSwitch().equals(targetFlow.getDestSwitch())) {
+            throw new InvalidFlowException("Couldn't add one-switch flow into affinity group",
+                    ErrorType.PARAMETERS_INVALID);
+        }
+
+        Flow affinityFlow = flowRepository.findById(targetFlow.getAffinityFlowId())
+                .orElseThrow(() ->
+                        new InvalidFlowException(format("Failed to find affinity flow id %s",
+                                targetFlow.getDiverseFlowId()), ErrorType.PARAMETERS_INVALID));
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/validation/FlowValidator.java]: Fixed.

File: src-java/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java [hunk 15]
```diff
@@ -99,6 +100,24 @@ public AvailableNetwork getAvailableNetwork(Flow flow, Collection<PathId> reuseP
                             }));
         }
 
+        // The main flow id is set as the affinity group id.
+        // All flow paths in this group are built along the path of the main flow.
+        // The main flow should not take into account the rest of the flows in this affinity group.
+        if (flow.getAffinityGroupId() != null && !flow.getAffinityGroupId().equals(flow.getFlowId())) {
+            log.info("Filling AvailableNetwork affinity weighs for group with id {}", flow.getAffinityGroupId());
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java]: Fixed.

File: src-java/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java [hunk 16]
```diff
@@ -81,10 +82,10 @@ public AvailableNetwork getAvailableNetwork(Flow flow, Collection<PathId> reuseP
             throw new RecoverableException("An error from the database", e);
         }
 
-        if (flow.getGroupId() != null) {
-            log.info("Filling AvailableNetwork diverse weighs for group with id {}", flow.getGroupId());
+        if (flow.getDiverseGroupId() != null) {
+            log.info("Filling AvailableNetwork diverse weighs for group with id {}", flow.getDiverseGroupId());
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java]: Fixed.

File: src-java/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java [hunk 17]
```diff
@@ -99,6 +100,24 @@ public AvailableNetwork getAvailableNetwork(Flow flow, Collection<PathId> reuseP
...
+        if (flow.getAffinityGroupId() != null && !flow.getAffinityGroupId().equals(flow.getFlowId())) {
+            log.info("Filling AvailableNetwork affinity weighs for group with id {}", flow.getAffinityGroupId());
+
+            Collection<FlowPath> flowPaths = flowPathRepository.findByFlowId(flow.getAffinityGroupId());
+
+            FlowPath foundFlowPath = flowPaths.stream()
+                    .filter(flowPath -> flowPath.getSrcSwitchId().equals(flow.getSrcSwitchId()))
+                    .findAny().orElse(null);
+            if (foundFlowPath != null) {
+                network.processAffinitySegments(foundFlowPath.getSegments());
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-pce/src/main/java/org/openkilda/pce/AvailableNetworkFactory.java:115]: Added unit test where approach of using all of the paths does not work.

File: src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/AvailableNetwork.java [hunk 18]
```diff
@@ -165,7 +169,58 @@ public void processDiversitySegmentsWithPop(List<PathSegment> segments) {
...
+    public void processAffinitySegments(List<PathSegment> segments) {
+        Set<Edge> pathEdges = new HashSet<>();
+        List<Node> pathNodes = new ArrayList<>();
+        for (int i = segments.size() - 1; i >= 0; i--) {
+            PathSegment segment = segments.get(i);
+            Node srcNode = getSwitch(segment.getSrcSwitchId());
+            Node dstNode = getSwitch(segment.getDestSwitchId());
+
+            if (srcNode == null || dstNode == null) {
+                log.debug("Affinity segment {} don't present in AvailableNetwork", segment);
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/AvailableNetwork.java]: Fixed

**Review by rozdy — APPROVED** [review]

**Review by sergii-iakovenko — APPROVED** [review]

**Pull request merged**